### ISSUE TITLE
Completion of error handling

### ISSUE THREAD

**Issue by elfring** [NLnetLabs/unbound#852]
Would you like to add more error handling for return values from functions like the following?
* [printf](https://pubs.opengroup.org/onlinepubs/9699919799/functions/printf.html "Print formatted output.") ⇒ [usage](https://github.com/NLnetLabs/unbound/blob/785c938106dbbfb2d5d93d4ee49cf4d34c04dcf5/smallapp/unbound-control.c#L89)
* [strdup](https://pubs.opengroup.org/onlinepubs/9699919799/functions/strdup.html "Duplicate a string.") ⇒ [readup_str](https://github.com/NLnetLabs/unbound/blob/a64cbe958dacb2cafe2800587fd76154371a464b/testcode/lock_verify.c#L163)

**Comment by wcawijngaards** [comment]
The strdup call is nicer with error handling. It is added in the commit above. The prints in the usage printout routine from unbound-control do not need such error handling, I would think.

Thanks for the heads up on the error oversight in the test program! That is nice to have a fix for.

**Comment by elfring** [comment]
I suggest to avoid [ignorance of return values](https://cwe.mitre.org/data/definitions/252.html "Checking more return values") a bit more.
Would you like to detect every error situation as early as possible?

**Comment by wcawijngaards** [comment]
I think it is a good idea to avoid ignorance of return values. But for such print statements, in the usage output, I think that is not nicer to check them.

There are a number of `(void)` casts that are used when return values are not inspected, already in the code. So that coverage can be extended. The value of this is debatable, it turns maybe into too many false positives and clutter. Or maybe it is useful as an annotation.

**Comment by elfring** [comment]
:thought_balloon: I propose to take also the possibility better into account for data output failures.

* :thinking: How do you think about to improve [static source code analysis](https://en.wikipedia.org/wiki/List_of_tools_for_static_code_analysis#C.2C_C.2B.2B "Tools for code analysis") also for this software?
* :crystal_ball: How will chances evolve to benefit any more also from [the means of aspect-oriented software development](https://en.wikipedia.org/wiki/Aspect-oriented_programming "Documentation")?

**Comment by wcawijngaards** [comment]
Yes, it is important to check for data output failures. For that, checking the return value of print function when NSD is printing to file, in other places, is important. But the usage error text data output is not significant.

The annotation, if useful for analysis, could prove useful. That could be a reason to add it.

Of course, we want the code to be as good as possible. I am not sure how this change would impact benefits from that programming paradigm.

**Comment by elfring** [comment]
:thinking: Can you become more aware that [error detection](https://en.wikipedia.org/wiki/Cross-cutting_concern#Examples "Documentation") belongs to known cross-cutting concerns?

**Comment by wcawijngaards** [comment]
Yes, that is true, that error handling fits in that way. That is interesting to read. :-)

For the usage printout, I think the result of handling the errors would make for annotated non-handling of it. Handling it to then try to printout another error that error printout does not work is not an improvement in program functionality for the usage printout, I think.

**Comment by elfring** [comment]
>That is interesting to read.

:eyes: [Some development approaches are worth for another look](https://www.aspectc.org/ "Software development with AspectC++"), aren't they?
